server/src/main/java/org/elasticsearch/search/query/TopDocsCollectorManagerFactory.java

```diff
@@ -1,527 +0,0 @@
-        private final Function<TopDocs, Float> maxScoreSupplier;
-
-        /**
-         * Ctr
-         * @param collapseContext The collapsing context
-         * @param sortAndFormats The query sort
-         * @param numHits The number of collapsed top hits to retrieve.
-         * @param trackMaxScore True if max score should be tracked
-         */
-        private CollapsingTopDocsCollectorManagerFactory(
-            CollapseContext collapseContext,
-            @Nullable SortAndFormats sortAndFormats,
-            int numHits,
-            boolean trackMaxScore,
-            @Nullable FieldDoc after
-        ) {
-            super(REASON_SEARCH_TOP_HITS, sortAndFormats == null ? new DocValueFormat[] { DocValueFormat.RAW } : sortAndFormats.formats);
-            assert numHits > 0;
-            assert collapseContext != null;
-            Sort sort = sortAndFormats == null ? Sort.RELEVANCE : sortAndFormats.sort;
-            this.topDocsCollector = collapseContext.createTopDocs(sort, numHits, after);
-
-            final MaxScoreCollector maxScoreCollector;
-            if (sortAndFormats == null) {
-                maxScoreCollector = null;
-                maxScoreSupplier = (topDocs) -> topDocs.scoreDocs.length == 0 ? Float.NaN : topDocs.scoreDocs[0].score;
-            } else if (trackMaxScore) {
-                maxScoreCollector = new MaxScoreCollector();
-                maxScoreSupplier = (topDocs) -> maxScoreCollector.getMaxScore();
-            } else {
-                maxScoreCollector = null;
-                maxScoreSupplier = (topDocs) -> Float.NaN;
-            }
-            this.collector = MultiCollector.wrap(topDocsCollector, maxScoreCollector);
-        }
-
-        @Override
-        Collector collector() {
-            return collector;
-        }
-
-        @Override
-        TopDocsAndMaxScore topDocsAndMaxScore() throws IOException {
-            TopFieldGroups topDocs = topDocsCollector.getTopGroups(0);
-            return new TopDocsAndMaxScore(topDocs, maxScoreSupplier.apply(topDocs));
-        }
-    }
-
-    static class SimpleTopDocsCollectorManagerFactory extends TopDocsCollectorManagerFactory {
-
-        private static TopDocsCollector<?> createCollector(
-            @Nullable SortAndFormats sortAndFormats,
-            int numHits,
-            @Nullable ScoreDoc searchAfter,
-            int hitCountThreshold
-        ) {
-            if (sortAndFormats == null) {
-                return TopScoreDocCollector.create(numHits, searchAfter, hitCountThreshold);
-            } else {
-                return TopFieldCollector.create(sortAndFormats.sort, numHits, (FieldDoc) searchAfter, hitCountThreshold);
-            }
-        }
-
-        private final Collector collector;
-        private final Supplier<TotalHits> totalHitsSupplier;
-        private final Supplier<TopDocs> topDocsSupplier;
-        private final Supplier<Float> maxScoreSupplier;
```

Comment:
These three suppliers are no longer needed, which feels great. They were used to set the right values within TopDocsAndMaxScore, which they were coming from different places depending on input parameters. It was previously difficult to trace which collector was providing them and how, due to the wrapping in different places. Now, we can have different subclasses of QueryPhaseCollectorManager (similar to what we previously has in TopDocsCollectorManagerFactory) that provide values depending on their logic, and those values become automatically part of the end result that the collector manager reduces which the search method returns.

This is the main reason to refactor the query phase like this PR does, as part of introducing support for concurrent collection to it. It felt really bad to convert to concurrent collector manager and keep the suppliers around.